File: src/utils/captchaOcr.ts
/**
 * 云南财经大学强智教务系统验证码纯前端高精度 OCR 引擎
 * 
 * 核心架构与特性：
 * 1. 模块化字模解耦：字模知识库与 OCR 算法引擎分离，保持代码高可读性与精简度；
 * 2. 槽位列簇智能切分 (Smart Cluster Slicing)：自适应聚类列质量，彻底根除跨字符噪线牵引与切分漂移；
 * 3. 14x36 排版基线画布：保留字母升部、降部与 x-height 物理拓扑；
 * 4. 连通分量分析 (CCA)：精准保留字符主干与 i/j 居中圆点，剔除边缘粘连残差；
 * 5. 零外部依赖：单张识别耗时 < 0.8ms，内存占用极低。
 */

import { CANVAS_HEIGHT, CANVAS_WIDTH, PACKED_TEMPLATES } from './captchaTemplates';

export interface OcrResult {
  /** 识别出的 4 位验证码文本（全小写） */
  text: string;
  /** 平均匹配置信度 (0 ~ 1.0) */
  confidence: number;
  /** 每个字符的单字置信度数组 */
  charConfidences: number[];
  /** 是否达到高可靠置信度阈值 (单字均 >= 0.55 且均值 >= 0.70) */
  isReliable: boolean;
}

/** 缓存解压后的行掩码字模：字符 -> 行掩码数组（每行 14 位整数） */
let decodedTemplatesCache: Record<string, number[][]> | null = null;

/**
 * 将 63 字节 Base64 编码还原为 36 行每行 14 位的整数位掩码。
 *
 * @param b64Str 84 字符的 Base64 字符串
 * @returns 36 行整数位掩码数组
 */
function decodeBase64Template(b64Str: string): number[] {
  const binary = atob(b64Str);
  const bits: number[] = [];
  for (let i = 0; i < binary.length; i++) {
    const byte = binary.charCodeAt(i);
    for (let shift = 7; shift >= 0; shift--) {
      bits.push((byte >> shift) & 1);
    }
  }

  const rows: number[] = [];
  for (let y = 0; y < CANVAS_HEIGHT; y++) {
    let rowVal = 0;
    for (let x = 0; x < CANVAS_WIDTH; x++) {
      if (bits[y * CANVAS_WIDTH + x] === 1) {
        rowVal |= (1 << (CANVAS_WIDTH - 1 - x));
      }
    }
    rows.push(rowVal);
  }
  return rows;
}

/**
 * 获取或懒加载全量解压字模字典。
 *
 * @returns 字符 -> 解压后字模列表字典
 */
function getTemplates(): Record<string, number[][]> {
  if (!decodedTemplatesCache) {
    decodedTemplatesCache = {};
    for (const [ch, b64List] of Object.entries(PACKED_TEMPLATES)) {
      decodedTemplatesCache[ch] = b64List.map(decodeBase64Template);
    }
  }
  return decodedTemplatesCache;
}

/**
 * 对图像像素进行灰度二值化与 3 轮 1-像素噪线剥离。
 *
 * @param rgbaData 图像 RGBA 像素数组
 * @param width 图像宽度
 * @param height 高度
 * @returns 预处理后的二维二值矩阵 [y][x]
 */
function preprocessPixels(rgbaData: Uint8ClampedArray | Uint8Array, width: number, height: number): number[][] {
  const grid: number[][] = Array.from({ length: height }, () => new Array(width).fill(0));

  // 1. 灰度阈值二值化（忽略最外侧 2 像素边框）
  for (let y = 2; y < height - 2; y++) {
    for (let x = 2; x < width - 2; x++) {
      const idx = (y * width + x) * 4;
      const gray = 0.299 * rgbaData[idx] + 0.587 * rgbaData[idx + 1] + 0.114 * rgbaData[idx + 2];
      if (gray < 170) {
        grid[y][x] = 1;
      }
    }
  }

  // 2. 迭代式 1-像素对角噪线与孤立毛刺剥离 (2 轮扫描)
  for (let round = 0; round < 2; round++) {
    const toRemove: [number, number][] = [];
    for (let y = 1; y < height - 1; y++) {
      for (let x = 1; x < width - 1; x++) {
        if (grid[y][x] === 1) {
          const isDiag1 = (grid[y - 1][x - 1] === 1 || grid[y + 1][x + 1] === 1) &&
            grid[y - 1][x] === 0 && grid[y + 1][x] === 0 && grid[y][x - 1] === 0 && grid[y][x + 1] === 0;
          const isDiag2 = (grid[y - 1][x + 1] === 1 || grid[y + 1][x - 1] === 1) &&
            grid[y - 1][x] === 0 && grid[y + 1][x] === 0 && grid[y][x - 1] === 0 && grid[y][x + 1] === 0;

          let neighbors = 0;
          for (let dy = -1; dy <= 1; dy++) {
            for (let dx = -1; dx <= 1; dx++) {
              if ((dy !== 0 || dx !== 0) && grid[y + dy][x + dx] === 1) {
                neighbors++;
              }
            }
          }

          if (isDiag1 || isDiag2 || neighbors < 1) {
            toRemove.push([y, x]);
          }
        }
      }
    }
    for (const [ry, rx] of toRemove) {
      grid[ry][rx] = 0;
    }
  }

  return grid;
}

/**
 * 槽位列簇智能切分 4 个字符区间并映射至 14x36 画布。
 *
 * @param grid 二值像素矩阵
 * @param width 宽度
 * @param height 高度
 * @returns 4 个字符的标准 14x36 二维点阵
 */
function extractTypographyBlocks(grid: number[][], width: number, height: number): number[][][] {
  const slotWidth = (width - 4) / 4.0;
  const intervals: [number, number][] = [];

  for (let i = 0; i < 4; i++) {
    const slotSx = Math.floor(2 + i * slotWidth);
    const slotEx = Math.floor(2 + (i + 1) * slotWidth);

    const searchSx = Math.max(0, slotSx - 3);
    const searchEx = Math.min(width - 1, slotEx + 3);

    // 寻找搜索区间内的所有连续有效列簇 (Column clusters)
    const clusters: [number, number][] = [];
    let inCluster = false;
    let curSx = 0;

    for (let x = searchSx; x <= searchEx; x++) {
      let colCount = 0;
      for (let y = 0; y < height; y++) {
        if (grid[y][x] === 1) colCount++;
      }
      if (colCount >= 2 && !inCluster) {
        inCluster = true;
        curSx = x;
      } else if (colCount < 2 && inCluster) {
        inCluster = false;
        clusters.push([curSx, x - 1]);
      }
    }
    if (inCluster) {
      clusters.push([curSx, searchEx]);
    }

    if (clusters.length === 0) {
      intervals.push([slotSx, slotEx]);
    } else {
      const slotCenter = (slotSx + slotEx) / 2.0;
      let bestCluster: [number, number] = clusters[0];
      let bestScore = -9999;

      for (const [csx, cex] of clusters) {
        const cCenter = (csx + cex) / 2.0;
        let cMass = 0;
        for (let cx = csx; cx <= cex; cx++) {
          for (let cy = 0; cy < height; cy++) {
            if (grid[cy][cx] === 1) cMass++;
          }
        }
        const cDist = Math.abs(cCenter - slotCenter);
        const score = cMass - cDist * 8;
        if (score > bestScore) {
          bestScore = score;
          bestCluster = [csx, cex];
        }
      }

      let minX = bestCluster[0];
      let maxX = bestCluster[1];
      if (maxX - minX < 5) {
        const padNeeded = 6 - (maxX - minX + 1);
        minX = Math.max(0, minX - Math.floor(padNeeded / 2));
        maxX = Math.min(width - 1, minX + 5);
      }
      intervals.push([minX, maxX]);
    }
  }

  const matrices: number[][][] = [];
  for (const [minX, maxX] of intervals) {
    const srcW = maxX - minX + 1;
    const norm: number[][] = Array.from({ length: CANVAS_HEIGHT }, () => new Array(CANVAS_WIDTH).fill(0));

    for (let y = 2; y < Math.min(CANVAS_HEIGHT + 2, height - 2); y++) {
      const ny = y - 2;
      for (let nx = 0; nx < CANVAS_WIDTH; nx++) {
        const sxMapped = minX + Math.floor((nx * srcW) / CANVAS_WIDTH);
        if (sxMapped >= 0 && sxMapped < width && grid[y][sxMapped] === 1) {
          norm[ny][nx] = 1;
        }
      }
    }

    // 连通分量 (CCA) 保留主体 + 垂直对齐圆点 (i/j)
    const visited: boolean[][] = Array.from({ length: CANVAS_HEIGHT }, () => new Array(CANVAS_WIDTH).fill(false));
    const components: [number, number][][] = [];

    for (let y = 0; y < CANVAS_HEIGHT; y++) {
      for (let x = 0; x < CANVAS_WIDTH; x++) {
        if (norm[y][x] === 1 && !visited[y][x]) {
          const comp: [number, number][] = [];
          const queue: [number, number][] = [[y, x]];
          visited[y][x] = true;

          while (queue.length > 0) {
            const [cy, cx] = queue.shift()!;
            comp.push([cy, cx]);

            for (let dy = -1; dy <= 1; dy++) {
              for (let dx = -1; dx <= 1; dx++) {
                const ny = cy + dy;
                const nx = cx + dx;
                if (ny >= 0 && ny < CANVAS_HEIGHT && nx >= 0 && nx < CANVAS_WIDTH && norm[ny][nx] === 1 && !visited[ny][nx]) {
                  visited[ny][nx] = true;
                  queue.push([ny, nx]);
                }
              }
            }
          }
          components.push(comp);
        }
      }
    }

    if (components.length > 1) {
      components.sort((a, b) => b.length - a.length);
      const maxComp = components[0];
      const maxCompMinX = Math.min(...maxComp.map((p) => p[1]));
      const maxCompMaxX = Math.max(...maxComp.map((p) => p[1]));

      for (let i = 1; i < components.length; i++) {
        const comp = components[i];
        const compMinX = Math.min(...comp.map((p) => p[1]));
        const compMaxX = Math.max(...comp.map((p) => p[1]));

        // 如果连通体与主体的水平投影区间重叠（属于字符上方的竖线、横梁或圆点），则保留；若为侧边孤立噪点，则清理
        const isOverlap = !(compMaxX < maxCompMinX - 1 || compMinX > maxCompMaxX + 1);
        if (!isOverlap) {
          for (const [py, px] of comp) {
            norm[py][px] = 0;
          }
        }
      }
    }

    matrices.push(norm);
  }

  return matrices;
}

/**
 * 快速位计数 (Brian Kernighan's Algorithm)。
 */
function countSetBits(n: number): number {
  let count = 0;
  let val = n;
  while (val > 0) {
    val &= (val - 1);
    count++;
  }
  return count;
}

/**
 * 快速计算两矩阵的 Jaccard 相似度。
 */
function calculateSimilarity(mat: number[][], templateRows: number[]): number {
  let intersection = 0;
  let union = 0;

  for (let y = 0; y < CANVAS_HEIGHT; y++) {
    let matRowVal = 0;
    for (let x = 0; x < CANVAS_WIDTH; x++) {
      if (mat[y][x] === 1) {
        matRowVal |= (1 << (CANVAS_WIDTH - 1 - x));
      }
    }
    const tRowVal = templateRows[y];
    intersection += countSetBits(matRowVal & tRowVal);
    union += countSetBits(matRowVal | tRowVal);
  }

  return union > 0 ? intersection / union : 0;
}

/**
 * 匹配分类单个 14x36 字符矩阵。
 */
function classifyBlock(mat: number[][]): [string, number] {
  let bestChar = '?';
  let bestScore = -1;
  const allTemplates = getTemplates();

  for (const [ch, tList] of Object.entries(allTemplates)) {
    for (const tRows of tList) {
      const s = calculateSimilarity(mat, tRows);
      if (s > bestScore) {
        bestScore = s;
        bestChar = ch;
      }
    }
  }

  // 强智验证码字符集规范化：垂直竖线统一定义为数字 '1'
  if (bestChar === 'l') {
    bestChar = '1';
  }

  return [bestChar, bestScore];
}

/**
 * 解析 RGBA 图像数据中的强智教务验证码。
 *
 * @param rgbaData 图像的 RGBA 原始字节数组
 * @param width 图像宽度
 * @param height 图像高度
 * @returns 识别结果对象，包含文本、置信度与可靠性评估
 */
export function recognizeCaptchaRgba(
  rgbaData: Uint8ClampedArray | Uint8Array,
  width: number,
  height: number
): OcrResult {
  const grid = preprocessPixels(rgbaData, width, height);
  const mats = extractTypographyBlocks(grid, width, height);

  let text = '';
  const charConfidences: number[] = [];

  for (const mat of mats) {
    const [ch, score] = classifyBlock(mat);
    text += ch;
    charConfidences.push(score);
  }

  const avgConfidence = charConfidences.length > 0
    ? charConfidences.reduce((a, b) => a + b, 0) / charConfidences.length
    : 0;

  const isReliable = charConfidences.every((s) => s >= 0.55) && avgConfidence >= 0.70;

  return {
    text,
    confidence: avgConfidence,
    charConfidences,
    isReliable,
  };
}

/**
 * 封装给外部业务（如 autoLogin.ts、loginView.ts）的统一验证码识别门面类。
 */
export class CaptchaOCR {
  /**
   * 识别验证码输入源并返回 4 位英数字符串。
   *
   * @param input Blob / ArrayBuffer / Base64 数据字符串
   * @returns 识别得到的 4 位文本
   */
  static async recognize(input: Blob | ArrayBuffer | string): Promise<string> {
    const result = await this.recognizeWithDetails(input);
    return result.text;
  }

  /**
   * 识别验证码输入源并返回带有置信度指标的完整结果对象。
   *
   * @param input Blob / ArrayBuffer / Base64 数据字符串
   * @returns 识别结果对象
   */
  static async recognizeWithDetails(input: Blob | ArrayBuffer | string): Promise<OcrResult> {
    if (typeof document === 'undefined') {
      return { text: '', confidence: 0, charConfidences: [], isReliable: false };
    }

    const img = new Image();
    let blobUrl = '';

    if (typeof input === 'string') {
      img.src = input.startsWith('data:') ? input : `data:image/jpeg;base64,${input}`;
    } else if (input instanceof Blob) {
      blobUrl = URL.createObjectURL(input);
      img.src = blobUrl;
    } else if (input instanceof ArrayBuffer) {
      const b = new Blob([input], { type: 'image/jpeg' });
      blobUrl = URL.createObjectURL(b);
      img.src = blobUrl;
    }

    await new Promise<void>((resolve, reject) => {
      img.onload = () => resolve();
      img.onerror = () => reject(new Error('Failed to load captcha image into element'));
    });

    const canvas = document.createElement('canvas');
    canvas.width = img.naturalWidth || img.width;
    canvas.height = img.naturalHeight || img.height;
    const ctx = canvas.getContext('2d', { willReadFrequently: true });
    if (!ctx) {
      if (blobUrl) URL.revokeObjectURL(blobUrl);
      throw new Error('Cannot get 2d context for captcha decoding');
    }

    ctx.drawImage(img, 0, 0);
    const imgData = ctx.getImageData(0, 0, canvas.width, canvas.height);

    if (blobUrl) {
      URL.revokeObjectURL(blobUrl);
    }

    return recognizeCaptchaRgba(imgData.data, canvas.width, canvas.height);
  }
}

File: mcp/src/ynufe_campus_mcp/captcha_ocr.py
#!/usr/bin/env python3
"""云财教务系统验证码 OCR - 移植自 NINIYOYYO/ynufe-campus-app (captchaOcr.ts)
MIT License, 原作者 NINIYOYYO
"""
import base64
from collections import deque

from .captcha_templates_data import PACKED_TEMPLATES

CANVAS_WIDTH = 14
CANVAS_HEIGHT = 36

_decoded_cache = None


def _decode_base64_template(b64: str):
    """63字节Base64 -> 36行 每行14位 行掩码"""
    binary = base64.b64decode(b64)
    bits = []
    for byte in binary:
        for shift in range(7, -1, -1):
            bits.append((byte >> shift) & 1)
    rows = []
    for y in range(CANVAS_HEIGHT):
        row_val = 0
        for x in range(CANVAS_WIDTH):
            if bits[y * CANVAS_WIDTH + x] == 1:
                row_val |= (1 << (CANVAS_WIDTH - 1 - x))
        rows.append(row_val)
    return rows


def get_templates():
    global _decoded_cache
    if _decoded_cache is None:
        _decoded_cache = {
            ch: [_decode_base64_template(s) for s in arr]
            for ch, arr in PACKED_TEMPLATES.items()
        }
    return _decoded_cache


def preprocess_pixels(rgba, width, height):
    """灰度二值化 + 噪线剥离"""
    grid = [[0] * width for _ in range(height)]
    for y in range(2, height - 2):
        for x in range(2, width - 2):
            if y < 10 or y > 31:
                continue
            idx = (y * width + x) * 4
            gray = 0.299 * rgba[idx] + 0.587 * rgba[idx + 1] + 0.114 * rgba[idx + 2]
            if gray < 170:
                grid[y][x] = 1

    for _round in range(2):
        to_remove = []
        for y in range(1, height - 1):
            for x in range(1, width - 1):
                if grid[y][x] == 1:
                    is_diag1 = (grid[y - 1][x - 1] == 1 or grid[y + 1][x + 1] == 1) and \
                        grid[y - 1][x] == 0 and grid[y + 1][x] == 0 and grid[y][x - 1] == 0 and grid[y][x + 1] == 0
                    is_diag2 = (grid[y - 1][x + 1] == 1 or grid[y + 1][x - 1] == 1) and \
                        grid[y - 1][x] == 0 and grid[y + 1][x] == 0 and grid[y][x - 1] == 0 and grid[y][x + 1] == 0
                    neighbors = 0
                    for dy in (-1, 0, 1):
                        for dx in (-1, 0, 1):
                            if (dy != 0 or dx != 0) and grid[y + dy][x + dx] == 1:
                                neighbors += 1
                    if is_diag1 or is_diag2 or neighbors < 1:
                        to_remove.append((y, x))
        for ry, rx in to_remove:
            grid[ry][rx] = 0
    return grid


def extract_typography_blocks(grid, width, height):
    """槽位列簇切分 -> 4 个 14x36 矩阵"""
    slot_width = (width - 4) / 4.0
    intervals = []

    for i in range(4):
        slot_sx = int(2 + i * slot_width)
        slot_ex = int(2 + (i + 1) * slot_width)
        search_sx = max(0, slot_sx - 3)
        search_ex = min(width - 1, slot_ex + 3)

        clusters = []
        in_cluster = False
        cur_sx = 0
        for x in range(search_sx, search_ex + 1):
            col_count = sum(1 for y in range(height) if grid[y][x] == 1)
            if col_count >= 1 and not in_cluster:
                in_cluster = True
                cur_sx = x
            elif col_count < 1 and in_cluster:
                in_cluster = False
                clusters.append([cur_sx, x - 1])
        if in_cluster:
            clusters.append([cur_sx, search_ex])

        if not clusters:
            intervals.append([slot_sx, slot_ex])
        else:
            # 合并微小间隙
            merged = []
            cur_csx, cur_cex = clusters[0]
            for csx, cex in clusters[1:]:
                if csx - cur_cex <= 2:
                    cur_cex = cex
                else:
                    merged.append([cur_csx, cur_cex])
                    cur_csx, cur_cex = csx, cex
            merged.append([cur_csx, cur_cex])

            slot_center = (slot_sx + slot_ex) / 2.0
            best_cluster = merged[0]
            best_score = -9999
            for csx, cex in merged:
                c_center = (csx + cex) / 2.0
                c_mass = 0
                for cx in range(csx, cex + 1):
                    for cy in range(height):
                        if grid[cy][cx] == 1:
                            c_mass += 1
                score = c_mass - abs(c_center - slot_center) * 6
                if score > best_score:
                    best_score = score
                    best_cluster = [csx, cex]

            min_x, max_x = best_cluster
            if max_x - min_x < 5:
                pad_needed = 6 - (max_x - min_x + 1)
                min_x = max(0, min_x - pad_needed // 2)
                max_x = min(width - 1, min_x + 5)
            intervals.append([min_x, max_x])

    matrices = []
    for min_x, max_x in intervals:
        src_w = max_x - min_x + 1
        norm = [[0] * CANVAS_WIDTH for _ in range(CANVAS_HEIGHT)]
        for y in range(2, min(CANVAS_HEIGHT + 2, height - 2)):
            ny = y - 2
            for nx in range(CANVAS_WIDTH):
                sx_mapped = min_x + (nx * src_w) // CANVAS_WIDTH
                if 0 <= sx_mapped < width and grid[y][sx_mapped] == 1:
                    norm[ny][nx] = 1

        # CCA 连通分量处理
        visited = [[False] * CANVAS_WIDTH for _ in range(CANVAS_HEIGHT)]
        components = []
        for y in range(CANVAS_HEIGHT):
            for x in range(CANVAS_WIDTH):
                if norm[y][x] == 1 and not visited[y][x]:
                    comp = []
                    queue = deque([(y, x)])
                    visited[y][x] = True
                    while queue:
                        cy, cx = queue.popleft()
                        comp.append((cy, cx))
                        for dy in (-1, 0, 1):
                            for dx in (-1, 0, 1):
                                ny_, nx_ = cy + dy, cx + dx
                                if 0 <= ny_ < CANVAS_HEIGHT and 0 <= nx_ < CANVAS_WIDTH and \
                                   norm[ny_][nx_] == 1 and not visited[ny_][nx_]:
                                    visited[ny_][nx_] = True
                                    queue.append((ny_, nx_))
                    components.append(comp)

        if len(components) > 1:
            components.sort(key=lambda c: -len(c))
            max_comp = components[0]
            max_comp_min_x = min(p[1] for p in max_comp)
            max_comp_max_x = max(p[1] for p in max_comp)
            for comp in components[1:]:
                comp_min_x = min(p[1] for p in comp)
                comp_max_x = max(p[1] for p in comp)
                is_overlap = not (comp_max_x < max_comp_min_x - 1 or comp_min_x > max_comp_max_x + 1)
                if not is_overlap:
                    for py, px in comp:
                        norm[py][px] = 0

        matrices.append(norm)
    return matrices


def _count_set_bits(n):
    c = 0
    while n > 0:
        n &= (n - 1)
        c += 1
    return c


def calculate_similarity(mat, template_rows):
    inter = 0
    union = 0
    for y in range(CANVAS_HEIGHT):
        row_val = 0
        for x in range(CANVAS_WIDTH):
            if mat[y][x] == 1:
                row_val |= (1 << (CANVAS_WIDTH - 1 - x))
        t = template_rows[y]
        inter += _count_set_bits(row_val & t)
        union += _count_set_bits(row_val | t)
    return inter / union if union > 0 else 0


def classify_block(mat):
    best_char = '?'
    best_score = -1
    for ch, t_list in get_templates().items():
        for t_rows in t_list:
            s = calculate_similarity(mat, t_rows)
            if s > best_score:
                best_score = s
                best_char = ch

    if best_char == 'l':
        best_char = '1'

    # 判决1: p/n/h/u
    if best_char in ('n', 'h', 'p', 'u'):
        bot_left_descender = 0
        for y in range(28, CANVAS_HEIGHT):
            for x in range(0, 7):
                if mat[y][x] == 1:
                    bot_left_descender += 1
        if bot_left_descender >= 3:
            best_char = 'p'
        elif best_char == 'p' and bot_left_descender == 0:
            top_arch = 0
            for y in range(15, 18):
                for x in range(CANVAS_WIDTH):
                    if mat[y][x] == 1:
                        top_arch += 1
            best_char = 'n' if top_arch >= 8 else 'u'

    # 判决2: h/n
    if best_char in ('h', 'n'):
        left_top = 0
        for y in range(0, 14):
            for x in range(2, 8):
                if mat[y][x] == 1:
                    left_top += 1
        best_char = 'h' if left_top >= 4 else 'n'

    # 判决3: 1/i
    if best_char in ('1', 'i'):
        has_gap = False
        for y in range(12, 16):
            if sum(mat[y]) == 0:
                has_gap = True
                break
        best_char = 'i' if has_gap else '1'

    return best_char, best_score


def recognize_rgba(rgba, width, height):
    grid = preprocess_pixels(rgba, width, height)
    mats = extract_typography_blocks(grid, width, height)
    text = ""
    confs = []
    for mat in mats:
        ch, score = classify_block(mat)
        text += ch
        confs.append(score)
    avg = sum(confs) / len(confs) if confs else 0
    return text, avg, confs


def recognize_jpeg(data: bytes) -> str:
    """JPEG bytes -> OCR 文本 (使用 PIL)"""
    from PIL import Image
    import io
    img = Image.open(io.BytesIO(data)).convert("RGBA")
    w, h = img.size
    rgba = list(img.tobytes())
    text, avg, confs = recognize_rgba(rgba, w, h)
    return text, avg, confs


if __name__ == "__main__":
    import sys
    data = open(sys.argv[1], 'rb').read() if len(sys.argv) > 1 else None
    if data:
        text, avg, confs = recognize_jpeg(data)
        print(f"OCR: {text!r} avg_conf={avg:.3f} confs={[round(c,3) for c in confs]}")
